File: core/runstate.go
package core

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"time"
)

// TaskStatus describes the lifecycle state of a task in a run.
type TaskStatus string

const (
	TaskStatusPending   TaskStatus = "pending"
	TaskStatusRunning   TaskStatus = "running"
	TaskStatusDone      TaskStatus = "done"
	TaskStatusFailed    TaskStatus = "failed"
	TaskStatusCancelled TaskStatus = "cancelled"
)

// TaskState records the result of a single task execution.
type TaskState struct {
	Status       TaskStatus `json:"status"`
	Output       string     `json:"output,omitempty"`
	Files        []string   `json:"files,omitempty"`
	CompletedAt  time.Time  `json:"completed_at,omitempty"`
	Error        string     `json:"error,omitempty"`
	PartialFiles []string   `json:"partial_files,omitempty"`
}

// RunState is the complete persisted state of a workflow run.
type RunState struct {
	ID           string                `json:"id"`
	WorkflowFile string                `json:"workflow_file"`
	Workflow     string                `json:"workflow"`
	Inputs       map[string]string     `json:"inputs"`
	Tasks        map[string]*TaskState `json:"tasks"`
}

// NewRunState creates a RunState for a new run.
func NewRunState(runID, workflowFile, workflowName string, inputs map[string]string) *RunState {
	return &RunState{
		ID:           runID,
		WorkflowFile: workflowFile,
		Workflow:     workflowName,
		Inputs:       inputs,
		Tasks:        make(map[string]*TaskState),
	}
}

// Save writes the RunState as JSON to <directory>/<runID>.json.
func Save(state *RunState, directory string) error {
	if err := os.MkdirAll(directory, 0755); err != nil {
		return fmt.Errorf("runstate: create directory: %w", err)
	}
	data, err := json.MarshalIndent(state, "", "  ")
	if err != nil {
		return fmt.Errorf("runstate: marshal: %w", err)
	}
	path := filepath.Join(directory, state.ID+".json")
	if err := os.WriteFile(path, data, 0644); err != nil {
		return fmt.Errorf("runstate: write %s: %w", path, err)
	}
	return nil
}

// Load reads a RunState from <directory>/<runID>.json.
func Load(runID, directory string) (*RunState, error) {
	path := filepath.Join(directory, runID+".json")
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("runstate: read %s: %w", path, err)
	}
	var state RunState
	if err := json.Unmarshal(data, &state); err != nil {
		return nil, fmt.Errorf("runstate: unmarshal %s: %w", path, err)
	}
	return &state, nil
}

// BuildRunState assembles a RunState from the outcome of a scheduler run.
// Tasks present in runContext.AllResults() are marked done; all others are
// marked cancelled when runErr is non-nil, or pending when it is nil.
func BuildRunState(runID string, workflowFile string, workflow *Workflow, inputs map[string]string, runContext *RunContext, runErr error) *RunState {
	state := NewRunState(runID, workflowFile, workflow.Name, inputs)
	completedResults := runContext.AllResults()

	for taskID := range workflow.Tasks {
		if result, ok := completedResults[taskID]; ok {
			state.Tasks[taskID] = &TaskState{
				Status: TaskStatusDone,
				Output: result.Output,
				Files:  result.Files,
			}
		} else if runErr != nil {
			state.Tasks[taskID] = &TaskState{Status: TaskStatusCancelled}
		} else {
			state.Tasks[taskID] = &TaskState{Status: TaskStatusPending}
		}
	}

	return state
}
func GenerateRunID() string {
	now := time.Now().UTC()
	suffix := fmt.Sprintf("%04x", now.UnixNano()%0xffff)
	return fmt.Sprintf("sh-%s-%s", now.Format("20060102"), suffix)
}

File: stagehand/core/runstate.py
from __future__ import annotations

import json
import os
import time
from dataclasses import dataclass, field
from typing import Optional


class TaskStatus:
    PENDING = "pending"
    RUNNING = "running"
    DONE = "done"
    FAILED = "failed"
    CANCELLED = "cancelled"
    SKIPPED = "skipped"


@dataclass
class TaskState:
    status: str = TaskStatus.PENDING
    output: str = ""
    files: list[str] = field(default_factory=list)
    completed_at: str = ""
    error: str = ""
    partial_files: list[str] = field(default_factory=list)

    def to_dict(self) -> dict:
        d: dict = {"status": self.status}
        if self.output:
            d["output"] = self.output
        if self.files:
            d["files"] = self.files
        if self.completed_at:
            d["completed_at"] = self.completed_at
        if self.error:
            d["error"] = self.error
        if self.partial_files:
            d["partial_files"] = self.partial_files
        return d

    @classmethod
    def from_dict(cls, data: dict) -> "TaskState":
        return cls(
            status=data.get("status", TaskStatus.PENDING),
            output=data.get("output", ""),
            files=data.get("files", []),
            completed_at=data.get("completed_at", ""),
            error=data.get("error", ""),
            partial_files=data.get("partial_files", []),
        )


@dataclass
class RunState:
    id: str
    workflow_file: str
    workflow: str
    inputs: dict[str, str] = field(default_factory=dict)
    tasks: dict[str, TaskState] = field(default_factory=dict)

    def to_dict(self) -> dict:
        return {
            "id": self.id,
            "workflow_file": self.workflow_file,
            "workflow": self.workflow,
            "inputs": self.inputs,
            "tasks": {k: v.to_dict() for k, v in self.tasks.items()},
        }

    @classmethod
    def from_dict(cls, data: dict) -> "RunState":
        tasks = {
            k: TaskState.from_dict(v) for k, v in data.get("tasks", {}).items()
        }
        return cls(
            id=data["id"],
            workflow_file=data.get("workflow_file", ""),
            workflow=data.get("workflow", ""),
            inputs=data.get("inputs", {}),
            tasks=tasks,
        )


def new_run_state(
    run_id: str,
    workflow_file: str,
    workflow_name: str,
    inputs: dict[str, str],
) -> RunState:
    return RunState(
        id=run_id,
        workflow_file=workflow_file,
        workflow=workflow_name,
        inputs=inputs,
    )


def save(state: RunState, directory: str) -> None:
    """Writes RunState as JSON to <directory>/<run_id>.json."""
    os.makedirs(directory, exist_ok=True)
    path = os.path.join(directory, state.id + ".json")
    with open(path, "w", encoding="utf-8") as f:
        json.dump(state.to_dict(), f, indent=2)


def load_state(run_id: str, directory: str) -> RunState:
    """Reads a RunState from <directory>/<run_id>.json."""
    path = os.path.join(directory, run_id + ".json")
    with open(path, "r", encoding="utf-8") as f:
        data = json.load(f)
    return RunState.from_dict(data)


def build_run_state(
    run_id: str,
    workflow_file: str,
    workflow: "Workflow",  # type: ignore[name-defined]
    inputs: dict[str, str],
    run_context: "RunContext",  # type: ignore[name-defined]
    run_err: Optional[Exception],
) -> RunState:
    """Assembles a RunState from the outcome of a scheduler run."""
    state = new_run_state(run_id, workflow_file, workflow.name, inputs)
    completed = run_context.all_results()
    skipped = run_context.skipped_ids()

    # Dynamically generated fan-out children (ids not in workflow.tasks) carry
    # their results in the run context; persist them so resume can skip them.
    dynamic_ids = [task_id for task_id in completed if task_id not in workflow.tasks]

    for task_id in list(workflow.tasks) + dynamic_ids:
        if task_id in skipped:
            state.tasks[task_id] = TaskState(status=TaskStatus.SKIPPED)
        elif task_id in completed:
            result = completed[task_id]
            state.tasks[task_id] = TaskState(
                status=TaskStatus.DONE,
                output=result.output,
                files=result.files,
            )
        elif run_err is not None:
            state.tasks[task_id] = TaskState(status=TaskStatus.CANCELLED)
        else:
            state.tasks[task_id] = TaskState(status=TaskStatus.PENDING)

    return state


def generate_run_id() -> str:
    now_ns = time.time_ns()
    from datetime import datetime, timezone
    now = datetime.now(timezone.utc)
    suffix = format(now_ns % 0xFFFF, "04x")
    return f"sh-{now.strftime('%Y%m%d')}-{suffix}"
